packages/casting/src/follower-cosmjs.js

```diff
@@ -112,290 +145,442 @@ export const makeCosmjsFollower = (
+    );
+    return result.slice(dataPrefixBytes.length);
+  };
 
-      if (result.length === 0) {
-        // No data.
-        return result;
+  /**
+   * @param {number} blockHeight
+   */
+  const tryGetDataAtHeight = blockHeight => {
+    const getProvenData = () => getProvenDataAtHeight(blockHeight);
+    const getUnprovenData = () => getUnprovenDataAtHeight(blockHeight);
+    return queryVerifier(getProvenData, crash, getUnprovenData);
+  };
+
+  /**
+   * @param {number} blockHeight
+   */
+  const getDataAtHeight = async blockHeight => {
+    for (let attempt = 0; ; attempt += 1) {
+      try {
+        // AWAIT
+        return await tryGetDataAtHeight(blockHeight);
+      } catch (error) {
+        // We expect occasionally to see an error here if the chain has not
+        // reached the requested blockHeight.
+        await E(leader).retry(where, error, attempt);
+        continue;
       }
+    }
+  };
 
-      // Handle the data prefix if any.
-      assert(
-        result.length >= dataPrefixBytes.length,
-        X`result too short for data prefix ${dataPrefixBytes}`,
-      );
-      assert(
-        dataPrefixBytes.every((v, i) => v === result[i]),
-        X`${result} doesn't start with data prefix ${dataPrefixBytes}`,
-      );
-      return result.slice(dataPrefixBytes.length);
+  /**
+   * @param {number} blockHeight
+   * @param {Uint8Array} data
+   */
+  const streamCellForData = (blockHeight, data) => {
+    const cell = decode(data);
+
+    if (
+      cell &&
+      Array.isArray(cell.values) &&
+      typeof cell.blockHeight === 'string'
+    ) {
+      blockHeight = Number(cell.blockHeight);
+      if (!Number.isNaN(blockHeight)) {
+        return {
+          blockHeight: Number(cell.blockHeight),
+          values: cell.values,
+        };
+      }
+    }
+
+    // Coerce legacy value cells to stream cells at their given height.
+    // Since this is either the first iteration or the data varies bytewise
+    // from the data on the previous block, we can assume the blockHeight is
+    // the current block.
+    return {
+      blockHeight,
+      values: [cell],
     };
+  };
+
+  /**
+   * @param {number} blockHeight
+   */
+  const getStreamCellAtHeight = async blockHeight => {
+    const data = await getDataAtHeight(blockHeight);
+    if (data.length !== 0) {
+      const streamCell = streamCellForData(blockHeight, data);
+      return { data, streamCell };
+    }
+    return undefined;
+  };
 
-  const getProvenValueAtHeight = makeQuerier('queryVerified');
-  const getUnprovenValueAtHeight = makeQuerier('queryUnverified');
+  /**
+   * @param {any} data
+   * @param {number} blockHeight
+   * @param {number} currentBlockHeight
+   * @returns {Promise<FollowerElement<T>>}
+   */
+  const followerElementFromStreamCellValue = async (
+    data,
+    blockHeight,
+    currentBlockHeight,
+  ) => {
+    if (typeof data === 'string') {
+      data = JSON.parse(data);
+    }
```

Comment:
Under what circumstances will `data` not be a string?
```suggestion
    assert.typeof(data, 'string');
    data = JSON.parse(data);
```

Reply:
Seems to be an artifact of the fakes for testing. I’ll drill and find the bug in the test scaffold.

Reply:
The fake RPC server did not JSON.stringify legacy value cells. That may have been by design, so I’m going to dig deeper. I might have to run an old chain with the new casting to verify.